File: generator/generator.ts
import path from 'path'
import type { TConfig, TContentItem, TMetadata } from './types'
import fs from 'fs'
import mergeImages from './merge_image'

const PATH_ROOT = path.join(__dirname, '..')

type TGenerator = {
    content: TContentItem[]
    generate: (count: number) => TContentItem[]
    drawImages: () => string
}

class RandomImagesGenerator implements TGenerator {
    private _content: TContentItem[] = []
    private _sequences: string[] = []
    private _imagesDirList: string [] = []
    private _maxSequencesCount: number = 0
    private _config: TConfig = null
    private _metadata: TMetadata = null

    private _choices(names: string[], weights: number[]): [string, string] {
        const sum = weights.reduce((prev, curr) => prev + curr, 0);
        let acc = 0;
        weights = weights.map(percent => (acc = percent + acc));        
        let randomNumber = Math.random() * sum;
        const element = names[weights.filter(percent => percent <= randomNumber).length]

        return [element, `${names.indexOf(element)}`]
    }

    private _calculateMaxSequenciesCount(): number {
        let totalSequenceAmount: number = 0
        for (const imageDir of this._imagesDirList) {
            const variantsCount = fs.readdirSync(path.join(PATH_ROOT, 'images', imageDir)).length
            totalSequenceAmount = !totalSequenceAmount ? variantsCount : totalSequenceAmount * variantsCount
        }

        return totalSequenceAmount
    }

    private _checkIsUniqueSequence(sequence: string): boolean {
        let isUnique: boolean = true
        for (const existesSequence of this._sequences) {
            if (existesSequence === sequence) {
                isUnique = false
                break
            }
        }

        return isUnique
    }

    constructor(imagesDirList: string[], metadata: TMetadata, config: TConfig) {
        this._imagesDirList = imagesDirList
        this._metadata = metadata
        this._config = config
        this._maxSequencesCount = this._calculateMaxSequenciesCount()
    }

    get content (): TContentItem[] { return this._content }

    generate(count: number = 1, sequencesIsUnique: boolean = true): TContentItem[] {
        const content: TContentItem[] = []
        
        while (count > 0) {
            if (this._sequences.length >= this._maxSequencesCount && sequencesIsUnique) {
                console.log(`The maximum possible number of sequences was generated: ${this._maxSequencesCount}`)
                break
            }

            const image: TContentItem = {}
            let sequence: string = ''

            for (let dirName of this._imagesDirList) {
                let contentList: string[] = fs.readdirSync(path.join(__dirname, '..', 'images', dirName))
                const names: string[] = []
                const percents: number[] = []

                for (const filename of contentList) {
                    let [ name, percent ]: string[] = filename.split('.')[0].split('#')
                    names.push(name)
                    percents.push(Number(percent))
                }

                let [ result, index ] = this._choices(names, percents)
                image[dirName] = result
                sequence += index
            }
            
            if (sequencesIsUnique) {
                if (this._checkIsUniqueSequence(sequence)) {
                    this._sequences.push(sequence)
                    content.push(image)
                    count--
                }
            } else {
                this._sequences.push(sequence)
                content.push(image)
                count--
            }
        }

        this._content = content
        return content
    }

    drawImages(): string {
        if (this._content.length == 0) {
            throw new Error(`You have not generate images: content = ${this._content.length}'`)
        }

        const outputDirPath = path.join(PATH_ROOT, 'assets')

        if (fs.existsSync(outputDirPath)) fs.rmSync(outputDirPath, { recursive: true })
        fs.mkdirSync(outputDirPath)

        if (!this._metadata) throw new Error('Token metadata cannot be empty.')

        let counter: number = 0

        const config = JSON.parse(JSON.stringify(this._config).replace('backgroud_color_rgba', 'backGroundColor'))
        for (const imageSettings of this._content) {

            mergeImages(`${counter}`, imageSettings, outputDirPath, config)

            let settings: TMetadata = {
                ...this._metadata,
                symbol: `${this._metadata.symbol}-${counter}`,
                name: `${this._metadata.name} #${counter}`,
                image: `${counter}.png`,
                properties: {
                    ...this._metadata.properties,
                    files: [{ uri: `${counter}.png`, type: 'image/png' }],
                },
                attributes: Object.entries(imageSettings).map(([ trait_type, value ]) => ({ trait_type, value }))
            }

            fs.writeFileSync(path.join(outputDirPath, `${counter}.json`), JSON.stringify(settings))
            counter++
        }

        return 'Success!'
    }
}

export default RandomImagesGenerator

File: server/images_generator/generator/generator.py
import random, os, shutil, json, time
from .merge_images import merge_images

PATH_ROOT = os.path.join(os.path.abspath(os.path.dirname(__file__)), '../../')

class RandomImageGenerator:
    __content = []
    __sequences = []
    __metadata = None
    __output_config = None

    def __choice(self, images, weights):
        result = random.choices(images, weights=weights)[0]
        return result, str(images.index(result))

    def __calculate_max_sequences_count(self):
        total_sequence_amount = 0
        for image_dir in self.images_dir_list:
            variants_count = len(os.listdir(os.path.join(PATH_ROOT, 'images', image_dir)))
            total_sequence_amount = total_sequence_amount * variants_count if total_sequence_amount != 0 else variants_count

        return total_sequence_amount

    def __init__(self, images_dir_list, token_metadata, output_config):
        self.images_dir_list = images_dir_list
        self.__metadata = token_metadata
        self.__output_config = output_config
        self.__max_sequences_count = self.__calculate_max_sequences_count()

    @property
    def content(self):
        return self.__content

    def generate(self, count = 1, sequences_is_unique = True, generator_time_limit = False):
        content = []
        timeout = time.time() + generator_time_limit if generator_time_limit != False else None

        while count > 0:
            if len(self.__sequences) >= self.__max_sequences_count and sequences_is_unique:
                print(f'The maximum possible number of sequences was generated: {self.__max_sequences_count}')
                break

            if timeout is not None and timeout - time.time() <= 0:
                break

            image = {}
            sequence = ''

            for dir_name in self.images_dir_list:
                content_list = os.listdir(os.path.join(PATH_ROOT, 'images', dir_name))
                names = []
                percents = []

                for filename in content_list:
                    name, percent = filename.split('.')[0].split('#')
                    names.append(name)
                    percents.append(int(percent))

                result, index = self.__choice(names, percents)
                image[dir_name] = result
                sequence += index

            if sequences_is_unique:
                if sequence not in self.__sequences:
                    self.__sequences.append(sequence)
                    content.append(image)
                    count -= 1
            else:
                self.__sequences.append(sequence)
                content.append(image)
                count -= 1

        self.__content = content
        return content

    def draw_images(self):
        if len(self.__content) == 0:
            raise ValueError(f'You have not generate images: len(self.content) = {len(self.__content)}')

        dist_dir = os.path.join(PATH_ROOT, 'assets')
        if os.path.exists(dist_dir) is True:
            shutil.rmtree(dist_dir)

        os.mkdir(dist_dir)

        counter = 0

        if self.__metadata is None:
            raise Exception(f'Missing token metadata. {os.path.join("root_configuration.json")}')

        for image_settings in self.__content:
            settings = {}
            merge_images(
                filename=counter,
                data=image_settings,
                output=dist_dir,
                conf=self.__output_config
            )

            for section, value in self.__metadata.items():
                settings[section] = value
                if section == 'symbol':
                    settings[section] = f'{value}-{counter}'

                if section == 'name':
                    settings[section] += f' #{counter}'

                if section == "image":
                    settings[section] = f'{counter}.png'

                if section == 'properties':
                    settings[section]["files"] = [{"uri": f'{counter}.png', "type": "image/png"}]

                if section == 'attributes':
                    settings[section] = [
                        {'trait_type': trait_type, 'value': value} for trait_type, value in image_settings.items()
                    ]
            try:
                with open(os.path.join(dist_dir, f'{counter}.json'), 'w') as fs:
                    json.dump(settings, fs, indent=4)
            except:
                raise Exception('Something went wrong while trying to write .json file for an image.')

            counter += 1

        return 'Success!'